books/xdoc/base.lisp

```diff
@@ -118,7 +120,18 @@
                           (cons :parents parents)
                           (cons :short short)
                           (cons :long long)
-                          (cons :from bookname)))
+                          ;; :from is the book where the doc is defined,
+                          ;; (usually relativized to acl2/books and with
+                          ;; :DIR :SYSTEM added)
+                          (cons :from bookname)
+                          ;; :code-from is specified by the xdoc writer,
+                          ;; and is not currently not modified/normalized
```

Comment:
not currently not => currently not

Reply:
I kind of think we ought to fix this before this enters widespread usage. It would seem strange to me to type in `:code-from "my/book/path :DIR :SYSTEM"` -- I suggest we allow both `:code-from "path"` and `:code-from ("path" :dir :dirname)`, normalize both to full-book-names using ACL2 system functions (parse-book-name, or something like it) and then use xdoc's normalize-bookname to get it into the form we want.

Reply:
That is a good idea, thanks!

Reply:
Let's explore this a little bit.  I like allowing `("path" :dir :system)` with `"path"` relative to `"acl2/books"`.

When you write `("path" :dir :dirname)` do you really mean `("path" :dir <project-keyword>)`?
If so, then I think we should interpret this `"path"` relative to the the project directory.
It would be confusing if `"path"` were relative to the current book containing the `defxdoc`.

Since the `:code-source` is supplied by the doc developer for doc purposes only, it is good to have clear rules on how the text is processed and what ends up in the rendered doc.  I think if you say
```
  :code-source ("something" :dir :myproject)
```
then it would be most clear if the `"something"` is passed unchanged (*) to the sysfile-p `(:myproject . "something")`, and then rendered in the html doc as 
```
  Code:  (include-book "something" :dir :myproject)
```

(*) There might be some minor modifications of adding and removing `".lisp"`, since `full-book-name`s are supposed to have the `".lisp"` extensions, but I am mainly referring to the directory of the `"something"` here.

I don't think that this makes code movement harder.  If a doc file is referring to a source file with a path relative to the doc file that ought to remain in the same relative place upon movement, then specify it as the string form, not the list form.

For the string version with a relative path
```
  :code-source "subdir/file"
```
the developer story becomes slightly more complex.  I think it is reasonable if we want this to be interpreted relative to the current book's path, since it saves the doc developer a lot of typing and makes it easier to move code.  If the current book is under a project directory, subdirectory "currdir", then this could still be rendered in the html doc as something like
```
  Code:  (include-book "currdir/subdir/file" :dir :myproject)
```

@solswords @ericwhitmansmith What do you think?

Reply:
This all seems good to me.  I think we should make the `:code-source` as easy as possible to write and maintain (and so support relative paths there).  When rendered, the `include-book` form should, whenever possible, be something that can be used from anywhere (and so should have a `:dir` option whenever possible).

Reply:
That sounds good. I could go either way (in the :dir-relative, `("something" :dir :myproject)` case) as to whether we do minimal processing and just produce a string directly from that object, or canonicalize using parse-book-name or something similar and then generate the string.  I think the main differences would be that the former might allow 1) totally bogus paths, and 2) paths relative to include-book-dirs as well as project dirs, I think 1) is disadvantageous. 2) could be nice except that include-book-dirs aren't global, so potentially a user might have to determine what cert.acl2 file (or whatever) to load to get that include-book-dir.  So I guess I'm leaning toward full processing and canonicalization with parse-book-name even for the :dir-relative case.